Jardilio route redirect fix

petebacondarwin opened the pull request:
See #3332

petebacondarwin committed:
refactor($route): move getting the template into its own function

petebacondarwin committed:
refactor($route): move resolving locals into its own function

petebacondarwin committed:
refactor($route): consolidate route locals processing into a single handler

petebacondarwin committed:
fix($route): don't process route change controllers and templates for `redirectTo` routes

If a route defines a `redirectTo` value, the current route should stop processing the route
and instead wait to process the route of the redirect.

Currently, what is happening is that Angular detects a redirectTo value and updates $location,
but then continues processing the route that was intended to be redirected.
Templates are loaded, resolves are processed, ng-view then updates the view with a new template
and instantiates the controller all for a route that should have been redirected.

A common use case for assigning a function to the redirectTo is to validation authentication,
permission check, or some other logic to determine if the route should continue or redirect
else where. In the end, this happens, but in between, unexpected results may occur by updating
the view and instantiating controllers for logic that wasn't expected to be executed.

http://plnkr.co/edit/8QlA0ouuePH3p35Ntmjy?p=preview

This commit checks for a url change after a potential redirect, it the url changed, and is
not `null` nor `undefined`, exit out and don't process current route change.

Closes #3332

BREAKING CHANGE

The $route service no longer instantiates controllers nor gets templates for routes
that have a `redirectTo` that is a defined value and different to the current route
url.

petebacondarwin commented:
I accidentally changed the authorship of the first commit. I will change back before merging...

gkalpak commented on test/ngRoute/routeSpec.js:
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-describe('$route', function() {
+fdescribe('$route', function() {
```
:scream:

petebacondarwin replied:
Oops

gkalpak commented on src/ngRoute/route.js:
```diff
@@ -588,46 +588,25 @@ function $RouteProvider() {
+              newUrl = nextRoute.redirectTo(nextRoute.pathParams, $location.path(), $location.search());
+              $location.url(newUrl).replace();
+            }
+            if (isDefined(newUrl) && url !== newUrl) {
```
I thought `null` would behave the same as `undefined`.

Narretz replied:
Should all falsy values abort the redirect?

petebacondarwin replied:
currently `null` and `""` are not treated like `undefined` and so this PR doesn't change that (to minimize BC at this stage).

gkalpak commented:
Where are the docs ? :stuck_out_tongue:

gkalpak commented:
I would also mention `resolve` amoong the things that don't get processed in the BC notice.

Narretz commented on test/ngRoute/routeSpec.js:
```diff
@@ -1061,6 +1061,55 @@ describe('$route', function() {
       });
     });
+
+    it('should not instantiate controller or process template for a redirected route', function() {
```
What about `resolve`?

petebacondarwin commented:
So at the moment null works differently so this pr does not change that. We should consider making null work like undefined in a follow up pr

petebacondarwin commented:
I'll add checks commit message and docs for resolve

petebacondarwin committed:
fix($route): don't process route change controllers and templates for `redirectTo` routes

If a route defines a `redirectTo` value, the current route should stop processing the route
and instead wait to process the route of the redirect.

Currently, what is happening is that Angular detects a redirectTo value and updates $location,
but then continues processing the route that was intended to be redirected.
Templates are loaded, resolves are processed, ng-view then updates the view with a new template
and instantiates the controller all for a route that should have been redirected.

A common use case for assigning a function to the redirectTo is to validation authentication,
permission check, or some other logic to determine if the route should continue or redirect
else where. In the end, this happens, but in between, unexpected results may occur by updating
the view and instantiating controllers for logic that wasn't expected to be executed.

http://plnkr.co/edit/8QlA0ouuePH3p35Ntmjy?p=preview

This commit checks for a url change after a potential redirect, it the url changed, and is
not `null` nor `undefined`, exit out and don't process current route change.

Closes #3332

BREAKING CHANGE

The $route service no longer instantiates controllers nor calls resolves or template functions
for routes that have a `redirectTo` unless the `redirectTo` is a function that returns
`undefined`.

petebacondarwin commented:
@gkalpak and @Narretz PTAL!

Narretz commented:
LGTM

Narretz commented:
Don't forget to change back the authorship though!

gkalpak commented on src/ngRoute/route.js:
```diff
@@ -136,6 +136,12 @@ function $RouteProvider() {
+   *      Routes that specify `redirectTo` will not have their controllers, template functions
+   *      or resolves called, the `$location` will be changed to the redirect url and route
+   *      processing will stop. The exception to this is if the `redirectTo` is a function that
+   *      returns `undefined`. In this case the route transition occurs as though their was no
```
_their_  -->  _there_

petebacondarwin replied:
Aghast!

gkalpak commented:
The commit message mentions that `null` cancels the redirection. Remove it when merging to avoid confusion.

LGTM

petebacondarwin commented:
Oops. Well I guess I have to fix the `null` path now then, eh?
